### ISSUE TITLE
[Bug] Crashes leave behind old tables

### ISSUE THREAD

**Issue by wyze** [0xOlias/ponder#220]
Crashes leave behind old tables and I am not sure which ones are safe to delete. Aside from trying to address the crashes, can we expose the current `versionId` used in the table names in the GraphQL interface so we can clean up old tables?

### Logs

```
indexer    - historical sync complete (server now responding as healthy)
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182881 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182882 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182883 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182884 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182885 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182886 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182887 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182888 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182889 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182890 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182891 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182917 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182918 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182919 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182920 with a minimum depth of 0
error      - WARNING: Deep reorg detected on arbitrum-goerli at block 26182921 with a minimum depth of 0
error      - Realtime sync queue error
Task:
  hash:              0xf80d9850627813cacca137fcb8870a433e5089e6ada7d55a7b7c178c03e71946
  parentHash:        0xc91effdc7527d48e5c528ce26a4ff03dcfa5f20f1f7379054e4a689d963b408a
  number:            0x18f85b1
  timestamp:         0x648b1266
  transactionCount:  3
Stack: TimeoutError: The request took too long to respond.

URL: https://goerli-rollup.arbitrum.io/rpc
Request body: {"method":"eth_getBlockByNumber","params":["0x18f8590",true]}

Details: The request timed out.
Version: viem@0.3.50
    at Object.http (/app/node_modules/.pnpm/viem@0.3.50_typescript@5.1.3/node_modules/viem/dist/cjs/utils/rpc.js:25:28)
    at request (/app/node_modules/.pnpm/viem@0.3.50_typescript@5.1.3/node_modules/viem/dist/cjs/clients/transports/http.js:19:55)
    at delay.count.count (/app/node_modules/.pnpm/viem@0.3.50_typescript@5.1.3/node_modules/viem/dist/cjs/utils/buildRequest.js:30:26)
    at attemptRetry (/app/node_modules/.pnpm/viem@0.3.50_typescript@5.1.3/node_modules/viem/dist/cjs/utils/promise/withRetry.js:15:36)
    at retry (/app/node_modules/.pnpm/viem@0.3.50_typescript@5.1.3/node_modules/viem/dist/cjs/utils/promise/withRetry.js:12:17)
    at runNextTicks (node:internal/process/task_queues:60:5)
    at listOnTimeout (node:internal/timers:538:9)
    at process.processTimers (node:internal/timers:512:7)

/app/node_modules/.pnpm/yoga-layout-prebuilt@1.10.0/node_modules/yoga-layout-prebuilt/yoga-layout/build/Release/nbind.js:53
        throw ex;
        ^

TimeoutError: The request took too long to respond.

URL: https://goerli-rollup.arbitrum.io/rpc
Request body: {"method":"eth_getBlockByNumber","params":["latest",true]}

Details: The request timed out.
Version: viem@0.3.50
    at Object.http (/app/node_modules/.pnpm/viem@0.3.50_typescript@5.1.3/node_modules/viem/dist/cjs/utils/rpc.js:25:28)
    at request (/app/node_modules/.pnpm/viem@0.3.50_typescript@5.1.3/node_modules/viem/dist/cjs/clients/transports/http.js:19:55)
    at delay.count.count (/app/node_modules/.pnpm/viem@0.3.50_typescript@5.1.3/node_modules/viem/dist/cjs/utils/buildRequest.js:30:26)
    at attemptRetry (/app/node_modules/.pnpm/viem@0.3.50_typescript@5.1.3/node_modules/viem/dist/cjs/utils/promise/withRetry.js:15:36)
    at retry (/app/node_modules/.pnpm/viem@0.3.50_typescript@5.1.3/node_modules/viem/dist/cjs/utils/promise/withRetry.js:12:17)
    at runNextTicks (node:internal/process/task_queues:60:5)
    at process.processTimers (node:internal/timers:509:9) {
  details: 'The request timed out.',
  docsPath: undefined,
  metaMessages: [
    'URL: https://goerli-rollup.arbitrum.io/rpc',
    'Request body: {"method":"eth_getBlockByNumber","params":["latest",true]}'
  ],
  shortMessage: 'The request took too long to respond.',
  version: 'viem@0.3.50'
}

Node.js v18.16.0
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 7.
```

**Comment by 0xOlias** [comment]
Couple questions to make sure I understand what's going on here:
1. The logs you included don't immediately look related to leftover/stale tables. Are they meant to be, or are the logs only related to the crashes?
2. Are you running this locally or in production? Against SQLite or Postgres?
3. What else can you share about the crashes you're seeing? From the logs, it seems like you might be getting rate limited by the RPC. Or perhaps you closed your laptop briefly?

The left behind tables are indeed quite annoying, would love to get to the bottom of it. For some context - Ponder is designed such that multiple instances can use to the same database simultaneously, so the "entity" tables need some mechanism to prevent name collisions. At the moment 

Now on to the crashes - when you run Ponder from the cli (`ponder dev`/`ponder start`), there is [an attempt](https://github.com/0xOlias/ponder/blob/main/packages/core/src/bin/ponder.ts#L46) to gracefully kill the instance on a SIGINT. The kill method [drops all of the now-stale entity tables](https://github.com/0xOlias/ponder/blob/main/packages/core/src/Ponder.ts#L256). There's a bug somewhere in here - any ideas?

We could expose the user store version ID so you can tell which ones are safe to delete, but I'd rather take a crack at resolving the crashes / leftover tables first.

**Comment by wyze** [comment]
> 1. The logs you included don't immediately look related to leftover/stale tables. Are they meant to be, or are the logs only related to the crashes?

Only related to the crashes. Not sure where I would pull leftover/stale table logs from.

> 2. Are you running this locally or in production? Against SQLite or Postgres?

Production using Postgres.

> 3. What else can you share about the crashes you're seeing? From the logs, it seems like you might be getting rate limited by the RPC. Or perhaps you closed your laptop briefly?

I've shared what I could find, but it's not running on my laptop, so wouldn't be that. Could be rate limited as it is a public RPC.

If the service crashes, there is no instance to teardown. By that time it's on the startup of the next process and doesn't seem to have a way to remember a previous `versionId`.

**Comment by 0xOlias** [comment]
Thanks. Out of curiousity, what cloud hosting provider are you using?

It might help to use a paid RPC provider plan. Ponder apps usually need one.

I just noticed you're using Arbitrum. I recently made a change to the realtime indexing strategy to gracefully handle reorgs, which I'm now realizing may have effectively broken Ponder on Arbitrum. 😮 

One of the tradeoffs with the new indexing approach was that Ponder would perform worse on chains with a very short block time. To better support Arbitrum and similar chains, it may make sense to offer another realtime indexing strategy that does not handle reorgs but sends a predictable # of RPC requests. Created an issue [here](https://github.com/0xOlias/ponder/issues/228), and will do some local testing to confirm this hypothesis soon.

Apologies for the instability!
